src/components/search/map/clusterer.tsx

```diff
@@ -0,0 +1,211 @@
+import {
+  createContext,
+  useContext,
+  useEffect,
+  useState,
+  Fragment,
+} from "react";
+import { MapContext } from "./map";
+import { Marker } from "./marker";
+import type {
+  MapContextType,
+  ClustererProps,
+  PinStoreType,
+  ClusterContextType,
+} from "./types";
+import type { Map } from "@yext/components-tsx-maps";
+import {
+  Unit,
+  Projection,
+  Coordinate,
+  GeoBounds,
+} from "@yext/components-tsx-geo";
+
+export const ClusterContext = createContext<ClusterContextType | null>(null);
+
+export function useClusterContext() {
+  const ctx = useContext(ClusterContext);
+
+  if (!ctx) {
+    throw new Error(
+      "Attempted to call useClusterContext() outside of <Clusterer>."
+    );
+  }
+
+  return ctx;
+}
+
+export const Clusterer = ({ children, ClusterTemplate }: ClustererProps) => {
+  const { map } = useContext(MapContext) as MapContextType;
+  const [pinStore, setPinStore] = useState<PinStoreType[]>([]);
+  const [clusters, setClusters] = useState<PinStoreType[][]>();
+  const [clustersToRender, setClustersToRender] = useState<JSX.Element[]>([]);
+
+  console.log(pinStore);
+
+  // Recalculate the clusters when either the pin store is updated or the map zoom level changes.
+  useEffect(() => {
+    setClusters(generateClusters(pinStore, map));
+  }, [pinStore, map.getZoom()]);
+
+  // When the clusters are updated, remove any pins in a cluster of more than 1 pin from the map.
+  // Then calculate the geo bounds of all the pins in the cluster and render a single marker
+  // at their center.
+  useEffect(() => {
+    if (clusters && clusters.length) {
```

Comment:
Nits but could use `clusters?.length`, and could avoid some nesting by flipping the condition and returning early